test/Find.lean

```diff
@@ -1,9 +1,186 @@
+Of these, 1 match your patterns.
+• non_linear_pattern_test1
+-/
+#guard_msgs in
+#find uniquenameforthistest, List.replicate ?n _ ++ List.replicate ?n _
+
+/--
+info: Found 2 definitions mentioning List.replicate, uniquenameforthistest and HAppend.hAppend.
+Of these, 2 match your patterns.
+• non_linear_pattern_test1
+• non_linear_pattern_test2
+-/
+#guard_msgs in
+#find uniquenameforthistest, List.replicate ?n _ ++ List.replicate ?m _
+
+/--
+info: Found 2 definitions mentioning List.replicate, uniquenameforthistest, Eq and HAppend.hAppend.
+Of these, 1 match your patterns.
+• non_linear_pattern_test1
+-/
+#guard_msgs in
+#find uniquenameforthistest, |- _ = List.replicate ?n _ ++ List.replicate ?m _
+
+/--
+info: Found 2 definitions mentioning List.replicate, uniquenameforthistest, Eq and HAppend.hAppend.
+Of these, 1 match your patterns.
+• non_linear_pattern_test2
+-/
+#guard_msgs in
+#find uniquenameforthistest, |- List.replicate ?n _ ++ List.replicate ?m _ = _
+
+theorem hyp_ordering_test1 {n : Nat} (_ : uniquenameforthistest = true) (h : 0 < n) :
+  0 ≤ n := Nat.le_of_lt h
+
+theorem hyp_ordering_test2 {n : Nat} (h : 0 < n) (_ : uniquenameforthistest = true) :
+  0 ≤ n := Nat.le_of_lt h
+
+/--
+info: Found 2 definitions mentioning LE.le, LT.lt, OfNat.ofNat, uniquenameforthistest and Eq.
+Of these, 2 match your patterns.
+• hyp_ordering_test1
+• hyp_ordering_test2
+-/
+#guard_msgs in
+#find ⊢ uniquenameforthistest = _ → 0 < ?n → _ ≤ ?n
+
+
+-- Regression test
+
+section LinearPatternTest
+
+class Star (R : Type _) where star : R → R
+export Star(star)
+
+/-- warning: declaration uses 'sorry' -/
+#guard_msgs in
+theorem star_comm_self' {R} [Mul R] [Star R] (x : R) : star x * x = x * star x := sorry
+
+/--
+info: Found 1 definitions mentioning Star.star.
+Of these, 1 match your patterns.
+• star_comm_self'
+-/
+#guard_msgs in
+#find star _
+
+/--
+info: Found 1 definitions mentioning HMul.hMul, Eq and Star.star.
+Of these, 1 match your patterns.
+• star_comm_self'
+-/
+#guard_msgs in
+#find star ?a * ?a = ?a * star ?_
+
+-- The following does not work for some strange reason
+
+/--
+info: Found 1 definitions mentioning HMul.hMul, Eq and Star.star.
+Of these, 1 match your patterns.
+• star_comm_self'
+-/
+#guard_msgs in
+#find star ?a * ?a = ?a * star ?a
+
+
+end LinearPatternTest
+
+section ListMapTest
+
+open Mathlib.Tactic.Find
+open Lean Elab Command
+
+elab s:"#assert_match " name_s:ident concl:(turnstyle)? query:term : command => liftTermElabM do
+    let pat ← Lean.Elab.Term.elabTerm query none
+    let name := Lean.TSyntax.getId name_s
+    let matcher ←
+      if concl.isSome
+      then matchConclusion pat
+      else matchAnywhere pat
+    let c := (← getEnv).constants.find! name
+    unless ← matcher c do
+      logErrorAt s "Pattern does not match when it should!"
+
+/-- error: Pattern does not match when it should! -/
+#guard_msgs in
+#assert_match List.map (?a -> ?b) -> List ?a -> List ?b
```

Comment:
There seems to be a bug in `matchConclusion` or nearby, maybe I should hold off merging until this is resolved. But I think I might need help.